### DISCUSSION TITLE
Seeking Guidance on Incorporating Back Stress into Crystal Plasticity Constitutive Model

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#25557] [Q&A Modules: Solid mechanics] [answered]
Dear Moose Experts,

I hope this message finds you well. I am currently working on the integration of back stress into a crystal plasticity constitutive model within the Moose framework. I have been referencing the work done by @ngrilli  in the [`c_pfor_am::FiniteStrainCrystalPlasticityBackstress` class](https://github.com/ngrilli/c_pfor_am/blame/main/src/materials/FiniteStrainCrystalPlasticityBackstress.C) to achieve this.

To provide some context, the[ `FiniteStrainCrystalPlasticityBackstress`](https://github.com/PengWei97/yinglong/commit/b1050829e09952cd23432b7312a3383c9cd3d713) class is constructed based on UserObjects classes. I am attempting to derive from `CrystalPlasticityKalidindiBackstress` to incorporate back stress into the the `CrystalPlasticityKalidindiUpdate` class. So far, I have overridden functions such as `calculateSlipRate()` and `calculateConstitutiveSlipDerivative()` to achieve this goal.

However, I am uncertain whether I should also make modifications to the `updateSubstepConstitutiveVariableValues` and `cacheStateVariablesBeforeUpdate` functions. Additionally, I would appreciate guidance on any other aspects that I might be overlooking in this process.

As part of my project, I need to modify the constitutive equations as follows:

1. For shear strain rate: 
$\dot{\gamma}^\alpha=\dot{\gamma}_o\left|\frac{\tau^\alpha-\chi^\alpha}{g^\alpha}\right|^M {sgn}\left(\tau^\alpha-\chi^\alpha\right)$

2. For back stress evolution: 
  $\(\dot{\chi}^\alpha = h \dot{\gamma}^\alpha - r \chi^\alpha\left|\dot{\gamma}^\alpha\right|\)$

I would be immensely grateful for any expert guidance or suggestions on how to effectively incorporate these changes and ensure the accurate implementation of the back stress concept into the crystal plasticity constitutive model within Moose.

Thank you in advance for your time and assistance. Your expertise is highly valued.

Best regards,

Wei Peng

**Comment by hugary1995** [comment]
https://github.com/idaholab/blackbear can already do this with the NEML plugin.

**Reply by PengWei97** [reply]
Thank you very much for providing me with another solution, but through the tips you gave me, I did not find a crystal plasticity model that considers back stress in [the NEML documentation](https://neml.readthedocs.io/en/dev/crystal_plasticity.html#). So, can you provide me with a more specific link or test example?

**Reply by hugary1995** [reply]
https://github.com/Argonne-National-Laboratory/neml/blob/dev/examples/cp/cpmodel.xml

**Reply by PengWei97** [reply]
Hi @hugary1995. Thank you for your prompt response and suggestions. Unfortunately, I couldn't find the required content in the link you provided. Additionally, when I explored the "Frederick-Armstrong" back stress evolution model in the NEML repository, I found that the documentation in `neml/doc/manual/manual.lyx` only contained the following information:

```bash
A Frederick-Armstrong model has
\begin_inset Formula $n=1$
\end_inset
```

Considering my goal is to incorporate the back stress term into the existing crystal plasticity model `CrystalPlasticityKalidindiUpdate`, as indicated in my initial equations, I am still seeking guidance on the implementation of this specific back stress model.

While I appreciate the alternative approach you suggested, it's worth mentioning that my ultimate objective is to couple the crystal plasticity module with a phase-field module. Hence, it seems that building upon the current Kalidindi model and adding an Armstrong-Frederick back stress term would be the most suitable choice.

I genuinely appreciate your assistance and insights on this matter. If you have any further suggestions or resources that could aid in this endeavor, please do share.

**Comment by PengWei97** [comment]
Dear @ngrilli, @sapitts, and @hugary1995,

I hope this message finds you well. I wanted to provide an update on my recent progress and the new problems I've encountered in my endeavor to incorporate backstress into the crystal plasticity model.

To begin with, I have successfully created the [`CrystalPlasticityKalidindiBackstress`](https://github.com/PengWei97/yinglong/blob/3c3364dc321ed182042d53a6491a9db788aca963/src/materials/crystal_plasticity/CrystalPlasticityKalidindiBackstress.C) class based on `CrystalPlasticityKalidindiUpdate`. Specifically, I added the backstress term `_tau_b` to the `calculateSlipRate()` and `calculateConstitutiveSlipDerivative()` functions. Additionally, I implemented the `updateBackstress()` function to calculate the evolution of back stress.

Following this, I created Case 1, where `c_backstress = 1.0e1` and `d_backstress = 1.0e-1`. Case 1 appears to compute without issues; however, it does generate some warnings to the terminal during each linear solve, such as:
```bash
*** Warning ***
The following warning occurred in the object "trial_xtalpl," of type "CrystalPlasticityKalidindiBackstress."

Maximum allowable slip increment exceeded 0.130827
```
Subsequently, I created Case 2 with `c_backstress = 1.0e2` and `d_backstress = 1.0`. However, during the calculation, it proceeds until `Time Step 12, time = 1.9333, dt = 1.90735e-07`, after which it fails to converge. Furthermore, at `Time Step 10`, it outputs "NAN" values to the terminal, as shown below:
```
*** Warning ***
Time Step 10, time = 1.93359, dt = 0.00078125
 0 Nonlinear |R| = 1.506118e+02
Nonlinear solve did not converge due to DIVERGED_FNORM_NAN iterations 0
 Solve Did NOT Converge!
Aborting as solve did not converge
```
In summary, while I have made progress in constructing `CrystalPlasticityKalidindiBackstress`, it still exhibits issues, and I am uncertain about the root cause. I believe your expertise in this area would be invaluable in resolving these challenges and ensuring the accurate implementation of the backstress concept into the crystal plasticity model.

I sincerely appreciate your time and assistance on this matter. Your guidance and insights would be greatly appreciated.

Best regards,

Wei Peng

**Reply by sapitts** [reply]
Hi @PengWei97,

I'm glad to see that you are using the `Update` based crystal plasticity classes because those are the classes we are currently maintaining. I am not familiar with @ngrilli's implementation nor the NEML implementation, so I will focus my response solely on the `CrystalPlasticityKalidindiUpdate`.

I would recommend inheriting your new `CrystalPlasticityKalidindiBackstress` class from the existing `CrystalPlasticityKalidindiUpdate` class. In the new `CrystalPlasticityKalidindiBackstress` class overwrite (to modify) the methods [calculateSlipRate](https://github.com/idaholab/moose/blob/e6741bceedf8fad608a2a5d048c2d1fb09afe7c8/modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdate.C#L91) and [calculateConstitutiveSlipDerivative](https://github.com/idaholab/moose/blob/e6741bceedf8fad608a2a5d048c2d1fb09afe7c8/modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdate.C#L127) to use the shear strain rate given by eqn 1 in your original post. 

Add a new method in [calculateStateVariableEvolutionRateComponent](https://github.com/idaholab/moose/blob/e6741bceedf8fad608a2a5d048c2d1fb09afe7c8/modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdate.C#L164) to calculate the back stress evolution with the `updateBackStress` method you mentioned in this latest post. (Make sure to also call the `CrystalPlasticityKalidindiUpdate::calculateStateVariableEvolutionRateComponent()` method first to keep that functionality in your new class.

You will also need to manage the total value updating and substep caching for the back stress variable; see the treatment of the forest dislocation variables in [CrystalPlasticityHCPDislocationSlipBeyerleinUpdate](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityHCPDislocationSlipBeyerleinUpdate.C). In particular pay attention to the following methods:

- [updateStateVariables](https://github.com/idaholab/moose/blob/e6741bceedf8fad608a2a5d048c2d1fb09afe7c8/modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityHCPDislocationSlipBeyerleinUpdate.C#L578)
- [areConstitutiveStateVariablesConverged](https://github.com/idaholab/moose/blob/e6741bceedf8fad608a2a5d048c2d1fb09afe7c8/modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityHCPDislocationSlipBeyerleinUpdate.C#L388)
- [setInitialConstitutiveVariableValues](https://github.com/idaholab/moose/blob/e6741bceedf8fad608a2a5d048c2d1fb09afe7c8/modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityHCPDislocationSlipBeyerleinUpdate.C#L313)
- [setSubstepConstitutiveVariableValues](https://github.com/idaholab/moose/blob/e6741bceedf8fad608a2a5d048c2d1fb09afe7c8/modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityHCPDislocationSlipBeyerleinUpdate.C#L326)
- [updateSubstepConstitutiveVariableValues](https://github.com/idaholab/moose/blob/e6741bceedf8fad608a2a5d048c2d1fb09afe7c8/modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityHCPDislocationSlipBeyerleinUpdate.C#L421)

The warning you copied about the maximum slip increment in your Case 1 may be related to the size of the timestep or the associated slip increment tolerance. The appearance of `NANs` in Case 2 may also indicated that the slip increment tolerance is too tight, or, more likely, that the incrementing of the back stress isn't handled properly. As a first step I would recommend adding print statements to monitor the back stress and slip increment values calculated by your code.

Hope this helps,
Stephanie

**Comment by ngrilli** [comment]
@PengWei97  the FiniteStrainCrystalPlasticityBackstress in my repository is based on the old monolithic code.

I have included the backstress with Armstrong-Frederick evolution also in a class derived from the new "stress_update_material_based" material class.

See here:
https://github.com/ngrilli/c_pfor_am/blob/main/src/materials/CrystalPlasticityDislocationUpdate.C

The backstress is simply updated in updateStateVariables()

I recommend you use this newer code.

**Reply by PengWei97** [reply, marked as answer]
Dear @sapitts and @ngrilli,

I trust this message finds you both well. I wanted to express my sincere gratitude for your invaluable assistance and guidance in addressing the challenges I encountered while incorporating backstress into the crystal plasticity model. With your help, I'm delighted to report that the issues mentioned earlier have been successfully resolved.

Taking into account the suggestions provided, as well as Ngrilli's reference to the `CrystalPlasticityDislocationUpdate`, I made the necessary modifications to the existing `CrystalPlasticityKalidindiBackstress` class. These adjustments have proven effective, and I have now achieved successful results in a uniaxial tension simulation. Below, you will find the results as well as the updated version of the [CrystalPlasticityKalidindiBackstress](https://github.com/PengWei97/yinglong/commit/fa4ccce318cc3d8ab1d714dc850bcd3ff5a9e5ae):

![image](https://github.com/idaholab/moose/assets/56183775/a4977cdf-13f7-4457-a571-65f49d9c2068)

Thank you once again for your invaluable help. I look forward to your continued guidance as we move forward.

Best regards,

wei peng

**Reply by sapitts** [reply]
Please consider submitting a PR with this class to the MOOSE repository--we'd be glad to help you through the process and even more glad to have a broader set of crystal plasticity capabilities for the wider crystal-plasticity-MOOSE community!

**Reply by PengWei97** [reply]
Dear @sapitts,

I want to express my sincere gratitude for extending the invitation to contribute the `CrystalPlasticityKalidindiBackstress` class to the MOOSE repository. I am truly honored and excited to be a part of this collaborative effort to enhance the crystal plasticity capabilities within the broader crystal-plasticity-MOOSE community.

As per your suggestion, I have taken the necessary steps to move `CrystalPlasticityKalidindiBackstress` to the moose-[backstressCPKa2023](https://github.com/PengWei97/moose/tree/backstressCPKa2023) branch. Furthermore, I have created an initial version of the `CrystalPlasticityKalidindiBackstress.md` documentation and relevant test cases in `exception_backstress.m`. These materials are admittedly a work in progress, and I'm committed to refining and expanding them to ensure clarity and usability.

I have also initiated a pull request (#25669) to facilitate the merging process, making this contribution available for review and further development.

Once again, I want to thank you and the entire MOOSE community for your warm welcome and support throughout this journey. I'm looking forward to collaborating closely and contributing to the growth of the MOOSE framework's crystal plasticity capabilities.

Please feel free to provide any further guidance or recommendations as we move forward. Your expertise and mentorship are highly appreciated.

Best regards,

wei peng
